### DISCUSSION TITLE
Can StreamRabbitListenerContainerFactory message listeners retry messages?

### DISCUSSION THREAD

**Discussion by daniel-bray-sonalake** [spring-projects/spring-amqp#1454] [Q&A] [answered]
Hi, thanks in advance for any help

We're using the `StreamRabbitListenerContainerFactory` to build our connection listener for our rabbit stream, and it's working fine, but there are some cases where we get timing issues where we know a retried message will eventually process ok.

Our issue is that, as far as I can tell, consumers created by this factory don't support retries

- This factory has an `setRetryTemplate` operation, and I've set it, but it seems to do nothing
- I've also tried setting an `errorHandler` in the `@RabbitListener` but it's not working.

I suspect the problem is that the issue here in [StreamMessageListenerAdapter#onStreamMessage](https://github.com/spring-projects/spring-amqp/blob/cac8fdc8feec5d0dff2bd94e6e02d204313290c9/spring-rabbit-stream/src/main/java/org/springframework/rabbit/stream/listener/adapter/StreamMessageListenerAdapter.java#L63) where, any exception is just swallowed, so no retry / error logic can be called.

So I'm wondering if this streams container factory just doesn't support retries: is this true?

Is there some other way to process streams that supports retries, and also follows the offset specification?

Thanks

**Comment by garyrussell** [comment, marked as answer]
The retry template in the listener container factory is for retrying publishing replies (for a request/reply listener).

```java
	/**
	 * Set a {@link RetryTemplate} to use when sending replies; added to each message
	 * listener adapter.
	 * @param retryTemplate the template.
	 * @see #setReplyRecoveryCallback(RecoveryCallback)
	 * @see AbstractAdaptableMessageListener#setRetryTemplate(RetryTemplate)
	 */
	public void setRetryTemplate(RetryTemplate retryTemplate) {
		this.retryTemplate = retryTemplate;
	}
```

We don't currently support request/reply for stream listeners.

Retries for deliveries in the classic listener containers is implemented using the advice chain (with a retry interceptor https://docs.spring.io/spring-amqp/docs/current/reference/html/#retry); there is currently no advice chain support in the `StreamListenerContainer`.

We can look into doing that; please open a new feature issue.

In the meantime, you can use a `RetryTemplate` within your listener method (or using spring-retry's `@Retryable` annotation on the listener should work too).

**Reply by garyrussell** [reply]
Here's an example:
```java
@SpringBootApplication
@EnableRetry
public class Rgh1454Application {

	private static final Logger log = LoggerFactory.getLogger(Rgh1454Application.class);

	public static void main(String[] args) {
		SpringApplication.run(Rgh1454Application.class, args);
	}

	@RabbitListener(queues = "fooStream")
	@Retryable(maxAttempts = 3, backoff = @Backoff(delay = 1_000, multiplier = 1.5, maxDelay = 10_000))
	public void listen(String in) {
	 	log.info(in);
		throw new RuntimeException("test");
	}

	@Bean
	ApplicationRunner runner(RabbitTemplate template) {
		return args -> {
			template.convertAndSend("fooStream", "bar");
		};
	}

}
```

**Reply by daniel-bray-sonalake** [reply]
> We don't currently support request/reply for stream listeners.

Thanks

> In the meantime, you can use a RetryTemplate within your listener method

Thanks, I'll check it out. 

> We can look into doing that; please open a new feature issue.

I'll do this

**Comment by daniel-bray-sonalake** [comment]
> We can look into doing that; please open a new feature issue.

Raised as: https://github.com/spring-projects/spring-amqp/issues/1455

**Reply by daniel-bray-sonalake** [reply]
> In the meantime, you can use a RetryTemplate within your listener method
> Thanks, I'll check it out.

Just to clarify - and with thanks - this will do as a workaround, but it's not a safe resolution, because if the spring app is restarted while we're retrying it this way, the message is lost. What we'd be hoping for here is the ability to throw an error so the context offset isn't updated
